File: 백준/Silver/4779. 칸토어 집합/칸토어 집합.py
def cantor(arr, start, N):
    if N <= 1:
        return

    N = N // 3
    K = N + start
    for i in range(K, K + N):
        arr[i] = ' '

    cantor(arr, start, N)
    cantor(arr, K + N, N)

while True:
    try:
        N = int(input())
        N_list = ['-' for _ in range(3**N)]
    except EOFError:
        break

    cantor(N_list, 0, len(N_list))
    print(''.join(N_list))

File: 백준/Silver/4779. 칸토어 집합/칸토어 집합.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        while (scanner.hasNextInt()) {
            int N = scanner.nextInt();
            int length = (int) Math.pow(3, N);
            char[] arr = new char[length];

            for (int i = 0; i < length; i++) {
                arr[i] = '-';
            }

            cantor(arr, 0, length);
            System.out.println(new String(arr));
        }

        scanner.close();
    }

    public static void cantor(char[] arr, int start, int length) {
        if (length <= 1) return;

        int third = length / 3;
        int midStart = start + third;

        for (int i = midStart; i < midStart + third; i++) {
            arr[i] = ' ';
        }

        cantor(arr, start, third);
        cantor(arr, midStart + third, third);
    }
}